test/SlowTests/Issues/RavenDB-19481.cs

```diff
@@ -0,0 +1,157 @@
+using Nest;
+using NetTopologySuite.Utilities;
+using Raven.Client;
+using Raven.Client.Documents.Operations;
+using Raven.Client.Documents.Operations.Backups;
+using Raven.Client.Documents.Session;
+using Raven.Server.Documents;
+using Raven.Server.ServerWide.Context;
+using Tests.Infrastructure;
+using Xunit;
+using Xunit.Abstractions;
+using static Raven.Server.Smuggler.Documents.CounterItem;
+using Assert = Xunit.Assert;
+
+namespace SlowTests.Issues
+{
+    public class RavenDB_19481 : RavenTestBase
+    {
+        public RavenDB_19481(ITestOutputHelper output) : base(output)
+        {
+        }
+
+
+        [RavenFact(RavenTestCategory.Smuggler)]
+        public async Task ShouldntReapplyClusterTransactionTwiceInRestore()
+        {
+            DoNotReuseServer();
+
+            var backupPath = NewDataPath(suffix: "BackupFolder");
+            using (var store = GetDocumentStore())
+            {
+                const string id = "users/1";
+
+                await RevisionsHelper.SetupRevisions(Server.ServerStore, store.Database);
+
+                using (var session = store.OpenAsyncSession(new SessionOptions
+                {
+                    TransactionMode = TransactionMode.ClusterWide
+                }))
+                {
+                    await session.StoreAsync(new User
+                    {
+                        Name = "Grisha"
+                    }, id);
+                    await session.SaveChangesAsync();
+                }
+
+                var stats = await store.Maintenance.SendAsync(new GetStatisticsOperation());
+                Assert.Equal(1, stats.CountOfRevisionDocuments);
+
+
+                var config = Backup.CreateBackupConfiguration(backupPath);
+                var backupTaskId = await Backup.UpdateConfigAndRunBackupAsync(Server, config, store);
+                await Backup.RunBackupAndReturnStatusAsync(Server, backupTaskId, store, isFullBackup: false, expectedEtag: 4);
+
+                var databaseName = $"restored_database-{Guid.NewGuid()}";
+
+                var clusterTransactions = new Dictionary<string, long>();
+                Server.ServerStore.ForTestingPurposesOnly().BeforeExecuteClusterTransactionBatch = (dbName, batch) =>
+                {
+                    if (dbName == databaseName)
+                    {
+                        lock (clusterTransactions) 
+                        {
+                            foreach (var clusterTx in batch)
+                            {
+                                var raftRequestId = clusterTx.Options.TaskId;
+
+                                if (clusterTransactions.ContainsKey(clusterTx.Options.TaskId) == false)
+                                    clusterTransactions.Add(raftRequestId, 1);
+                                else
+                                    clusterTransactions[raftRequestId]++;
+                            }
+                        }
+                    }
+                };
```

Comment:
Can't do it because for that you need to create the db before and the apply this on its 'TestingStuff' but you cant restore to an existing db:
```
Raven.Client.Exceptions.RavenException: System.ArgumentException: Cannot restore data to an existing database named restored_database-7f12ce2a-2b89-48ce-8e67-d127f4c5ec3a
   at Raven.Server.Documents.PeriodicBackup.Restore.RestoreBackupTaskBase..ctor(ServerStore serverStore, RestoreBackupConfigurationBase restoreFromConfiguration, String nodeTag, OperationCancelToken operationCancelToken) in D:\work\RavenDB-19481_3\sln\src\Raven.Server\Documents\PeriodicBackup\Restore\RestoreBackupTaskBase.cs:line 87
   at Raven.Server.Documents.PeriodicBackup.Restore.RestoreFromLocal..ctor(ServerStore serverStore, RestoreBackupConfiguration restoreConfiguration, String nodeTag, OperationCancelToken operationCancelToken) in D:\work\RavenDB-19481_3\sln\src\Raven.Server\Documents\PeriodicBackup\Restore\RestoreFromLocal.cs:line 17
   at Raven.Server.Web.System.AdminDatabasesHandler.RestoreDatabase() in D:\work\RavenDB-19481_3\sln\src\Raven.Server\Web\System\AdminDatabasesHandler.cs:line 618
   at Raven.Server.Routing.RequestRouter.HandlePath(RequestHandlerContext reqCtx) in D:\work\RavenDB-19481_3\sln\src\Raven.Server\Routing\RequestRouter.cs:line 412
   at Raven.Server.RavenServerStartup.RequestHandler(HttpContext context) in D:\work\RavenDB-19481_3\sln\src\Raven.Server\RavenServerStartup.cs:line 184
 ---> System.ArgumentException: System.ArgumentException: Cannot restore data to an existing database named restored_database-7f12ce2a-2b89-48ce-8e67-d127f4c5ec3a
   at Raven.Server.Documents.PeriodicBackup.Restore.RestoreBackupTaskBase..ctor(ServerStore serverStore, RestoreBackupConfigurationBase restoreFromConfiguration, String nodeTag, OperationCancelToken operationCancelToken) in D:\work\RavenDB-19481_3\sln\src\Raven.Server\Documents\PeriodicBackup\Restore\RestoreBackupTaskBase.cs:line 87
   at Raven.Server.Documents.PeriodicBackup.Restore.RestoreFromLocal..ctor(ServerStore serverStore, RestoreBackupConfiguration restoreConfiguration, String nodeTag, OperationCancelToken operationCancelToken) in D:\work\RavenDB-19481_3\sln\src\Raven.Server\Documents\PeriodicBackup\Restore\RestoreFromLocal.cs:line 17
   at Raven.Server.Web.System.AdminDatabasesHandler.RestoreDatabase() in D:\work\RavenDB-19481_3\sln\src\Raven.Server\Web\System\AdminDatabasesHandler.cs:line 618
   at Raven.Server.Routing.RequestRouter.HandlePath(RequestHandlerContext reqCtx) in D:\work\RavenDB-19481_3\sln\src\Raven.Server\Routing\RequestRouter.cs:line 412
   at Raven.Server.RavenServerStartup.RequestHandler(HttpContext context) in D:\work\RavenDB-19481_3\sln\src\Raven.Server\RavenServerStartup.cs:line 184
   --- End of inner exception stack trace ---
   at Raven.Client.Exceptions.ExceptionDispatcher.Throw(JsonOperationContext context, HttpResponseMessage response, Action`1 additionalErrorInfo) in D:\work\RavenDB-19481_3\sln\src\Raven.Client\Exceptions\ExceptionDispatcher.cs:line 119
   at Raven.Client.Http.RequestExecutor.HandleUnsuccessfulResponse[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, HttpRequestMessage request, HttpResponseMessage response, String url, SessionInfo sessionInfo, Boolean shouldRetry, CancellationToken token) in D:\work\RavenDB-19481_3\sln\src\Raven.Client\Http\RequestExecutor.cs:line 1577
   at Raven.Client.Http.RequestExecutor.ExecuteAsync[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, Boolean shouldRetry, SessionInfo sessionInfo, CancellationToken token) in D:\work\RavenDB-19481_3\sln\src\Raven.Client\Http\RequestExecutor.cs:line 969
   at Raven.Client.Http.RequestExecutor.ExecuteAsync[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, Boolean shouldRetry, SessionInfo sessionInfo, CancellationToken token) in D:\work\RavenDB-19481_3\sln\src\Raven.Client\Http\RequestExecutor.cs:line 995
   at Raven.Client.ServerWide.Operations.ServerOperationExecutor.SendAsync(IServerOperation`1 operation, CancellationToken token) in D:\work\RavenDB-19481_3\sln\src\Raven.Client\ServerWide\Operations\ServerOperationExecutor.cs:line 134
   at Raven.Client.Util.AsyncHelpers.RunSync[T](Func`1 task) in D:\work\RavenDB-19481_3\sln\src\Raven.Client\Util\AsyncHelpers.cs:line 91
   at Raven.Client.ServerWide.Operations.ServerOperationExecutor.Send(IServerOperation`1 operation) in D:\work\RavenDB-19481_3\sln\src\Raven.Client\ServerWide\Operations\ServerOperationExecutor.cs:line 125
   at FastTests.RavenTestBase.BackupTestBase.RestoreDatabase(IDocumentStore store, RestoreBackupConfiguration config, Nullable`1 timeout, String nodeTag) in D:\work\RavenDB-19481_3\sln\test\Tests.Infrastructure\RavenTestBase.Backup.cs:line 293
   at SlowTests.Issues.RavenDB_19481.ShouldntReapplyClusterTransactionTwiceInRestore() in D:\work\RavenDB-19481_3\sln\test\SlowTests\Issues\RavenDB-19481.cs:line 95
   at Tryouts.Program.Main(String[] args) in D:\work\RavenDB-19481_3\sln\test\Tryouts\Program.cs:line 35
```

Reply:
okay, make sense